File: 2015/solutions/day13.py
from itertools import permutations

with open('../inputs/input13.txt', 'r') as f:
    data = f.readlines()

    def find_best():
        paths = []
        people = len(adjacency) - 1
        for perm in list(permutations(range(people + 1))):
            happiness = 0
            for i in range(people):
                happiness += adjacency[perm[i]][perm[i + 1]]
                happiness += adjacency[perm[i + 1]][perm[i]]
            happiness += adjacency[perm[people]][perm[0]]
            happiness += adjacency[perm[0]][perm[people]]
            paths.append(happiness)
        return max(paths)

    # part one
    hash = {'Alice': 0,
            'Bob': 1,
            'Carol': 2,
            'David': 3,
            'Eric': 4,
            'Frank': 5,
            'George': 6,
            'Mallory': 7}

    adjacency = [[0] * 8 for i in range(8)]
    for line in data:
        parsed = line.split()
        val = int(parsed[3]) if parsed[2] == 'gain' else -int(parsed[3])
        adjacency[hash[parsed[0]]][hash[parsed[-1][:-1]]] = val
    print(find_best())

    # part two
    for row in adjacency:
        row.append(0)
    adjacency.append([0] * 8)
    print(find_best())

File: 2017/solutions/day13.js
const fs = require('fs');

let run_packet = (delay, firewall) => {
    let sum = 0;
    for (let i = 0; i < 99; i++) {
        if (firewall[i] !== undefined && (i + delay) % (2 * (firewall[i] - 1)) === 0) sum += i * firewall[i];
    }
    return sum;
}

fs.readFile('../inputs/input13.txt', 'utf-8', (err, data) => {
    data = data.trim().split('\n');
    let firewall = Array(99);
    for (let layer of data) {
        let depth, range;
        [depth, range] = layer.split(': ').map(x => parseInt(x));
        firewall[depth] = range;
    }

    //part one
    console.log(run_packet(0, firewall));

    //part two
    let delay = 0;
    while (run_packet(delay++, firewall) !== 0 || (delay - 1) % 4 === 0);
    console.log(delay - 1);
});